src/hyperv/vmbus/vmbus_et.c

```diff
@@ -0,0 +1,117 @@
+/*-
+ * Copyright (c) 2015,2016-2017 Microsoft Corp.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *      notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in the
+ *      documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <unix_internal.h>
+#include <hyperv_internal.h>
+#include "hyperv_reg.h"
+#include "hyperv_var.h"
+#include "vmbus_var.h"
+
+//#define VMBUS_TIMER_DEBUG
+#ifdef VMBUS_TIMER_DEBUG
+#define vmbus_timer_debug(x, ...) do { rprintf("VMBUS TIMER: " x, ##__VA_ARGS__); } while(0)
+#else
+#define vmbus_timer_debug(x, ...)
+#endif
+
+#define MSR_HV_STIMER0_CFG_SINT        \
+    ((((uint64_t)VMBUS_SINT_TIMER) << MSR_HV_STIMER_CFG_SINT_SHIFT) & \
+     MSR_HV_STIMER_CFG_SINT_MASK)
+
+/*
+ * Additionally required feature:
+ * - SynIC is needed for interrupt generation.
+ */
+#define CPUID_HV_ET_MASK        (CPUID_HV_MSR_SYNIC |        \
+                     CPUID_HV_MSR_SYNTIMER)
+
+struct vmbus_timer {
+} vmbus_timer;
+
+typedef struct vmbus_timer *vmbus_timer_t;
+
+static vmbus_timer_t vmbus_et;
+
+static __inline u64
+hyperv_sbintime2count(sbintime_t time)
+{
+    struct timespec ts;
+    timespec_from_time(&ts, time);
```

Comment:
This is pulling in a unix type just to extract seconds and nanoseconds from a timestamp. You can just use sec_from_timestamp(time) and nsec_from_timestamp(truncate_seconds(time)) instead.